tough/src/editor.rs

```diff
@@ -0,0 +1,269 @@
+// Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: MIT OR Apache-2.0
+
+use crate::datastore::Datastore;
+use crate::error::{self, Result};
+use crate::key_source::KeySource;
+use crate::root_digest::{RootDigest, RootKeys};
+use crate::schema::{
+    Hashes, Role, RoleType, Root, Signature, Signed, Snapshot, SnapshotMeta, Target, Targets,
+    Timestamp, TimestampMeta,
+};
+use crate::transport::{TargetMapTransport, Transport};
+use crate::{Repository, Settings};
+use maplit::hashmap;
+use olpc_cjson::CanonicalFormatter;
+use ring::digest::{digest, SHA256, SHA256_OUTPUT_LEN};
+use ring::rand::{SecureRandom, SystemRandom};
+use serde::Serialize;
+use snafu::ResultExt;
+use std::collections::HashMap;
+use std::num::NonZeroU64;
+use std::path::{Path, PathBuf};
+use url::Url;
+
+/// An in-memory, unsigned TUF repository
+///
+/// This is provided as means to create and update a repository, which can then be signed.
+#[derive(Debug)]
+pub struct RepositoryEditor {
+    pub snapshot: Snapshot,
+    pub timestamp: Timestamp,
+    pub targets: Targets,
+    targets_meta: TargetMapTransport,
+}
+
+impl RepositoryEditor {
+    pub fn new(snapshot: Snapshot, timestamp: Timestamp, targets: Targets) -> RepositoryEditor {
+        RepositoryEditor {
+            snapshot,
+            timestamp,
+            targets,
+            targets_meta: TargetMapTransport {
+                targets: HashMap::new(),
+            },
+        }
+    }
+
+    //pub fn from_path<P>(path: P)
+    //where
+    //    P: AsRef<Path>,
+    //{
+    //}
+
+    pub fn add_target<P>(&mut self, path: P) -> Result<()>
+    where
+        P: AsRef<Path>,
+    {
+        // Get the absolute path to the given target
+        let path = std::fs::canonicalize(path.as_ref()).context(error::AbsolutePath {
+            path: path.as_ref().to_owned(),
+        })?;
+
+        // Build a Target and add it to the Targets struct
+        let (target_name, target) =
+            Target::from_path(&path).context(error::TargetFromPath { path: &path })?;
+        self.targets
+            .targets
+            .insert(target_name.clone(), target.clone());
+
+        // Add the target metadata to the TargetMapTransport
+        // TODO: NOTE HERE ABOUT BOTH FORMS
+        let target_sha256 = &target.hashes.sha256.clone().into_vec();
+        self.targets_meta.targets.insert(
+            format!("{}.{}", hex::encode(target_sha256), target_name),
+            path.to_owned(),
+        );
+        self.targets_meta
+            .targets
+            .insert(target_name, path.to_owned());
+
+        Ok(())
+    }
+
+    pub fn build<'a, P1, P2, T>(
```

Comment:
I would probably break this into a couple of subroutines. For example where you have the comments and separators, those seem like good subroutine delineations for refactor.